plutarch-test/conditional/Plutarch/TryFromSpec.hs

```diff
@@ -0,0 +1,506 @@
+             )
+          @-> psucceeds
+        "[String] == [String]"
+          @| checkDeepUnwrap
+            @(PBuiltinList (PAsData PByteString))
+            @(PBuiltinList (PAsData PByteString))
+            (pdata $ (pcons # (pdata $ pconstant "foo")) #$ (psingleton # (pdata $ pconstant "bar")))
+          @-> psucceeds
+      "partial checks" @\ do
+        -- this is way more expensive ...
+        "check whole structure"
+          @| fullCheck @-> psucceeds
+        -- ... than this
+        "check structure partly"
+          @| partialCheck @-> psucceeds
+      "recovering a nested record" @\ do
+        "succeeds"
+          @| checkDeep
+            @(PDataRecord '["_0" ':= (PDataRecord '["_1" ':= PInteger])])
+            @(PDataRecord '["_0" ':= (PDataRecord '["_1" ':= PInteger])])
+            (pdata $ pdcons # (pdata $ pdcons # pdata (pconstant 42) # pdnil) # pdnil)
+          @-> psucceeds
+        "fails"
+          @| checkDeep
+            @(PDataRecord '["_0" ':= (PDataRecord '["_1" ':= PByteString])])
+            @(PDataRecord '["_0" ':= (PDataRecord '["_1" ':= PInteger])])
+            (pdata $ pdcons # (pdata $ pdcons # pdata (pconstant 42) # pdnil) # pdnil)
+          @-> pfails
+        "sample usage contains the right value"
+          @| pconstant 42 #== theField @-> passert
+    "example" @\ do
+      let validContext0 = ctx validOutputs0 validList1
+          invalidContext1 = ctx invalidOutputs1 validList1
+          l1 :: Term _ (PAsData (PBuiltinList (PAsData PInteger)))
+          l1 = toDatadList [1 .. 5]
+          l2 :: Term _ (PAsData (PBuiltinList (PAsData PInteger)))
+          l2 = toDatadList [6 .. 10]
+          l3 :: Term _ (PAsData (PBuiltinList (PAsData PInteger)))
+          l3 = toDatadList [6 .. 9]
+          l4 :: Term _ (PAsData (PBuiltinList (PAsData PInteger)))
+          l4 = toDatadList [6, 8, 8, 9, 10]
+      "concatenate two lists, legal"
+        @| validator # pforgetData l1 # pforgetData l2 # validContext0 @-> psucceeds
+      "concatenate two lists, illegal (list too short)"
+        @| validator # pforgetData l1 # pforgetData l3 # validContext0 @-> pfails
+      "concatenate two lists, illegal (wrong elements in list)"
+        @| validator # pforgetData l1 # pforgetData l4 # validContext0 @-> pfails
+      "concatenate two lists, illegal (more than one output)"
+        @| validator # pforgetData l1 # pforgetData l2 # invalidContext1 @-> pfails
+    "example2" @\ do
+      "recovering a record succeeds"
+        @| recoverAB @-> psucceeds
+
+------------------- Checking deeply, shallowly and unwrapping ----------------------
+
+checkDeep ::
+  forall (target :: PType) (actual :: PType).
+  ( PTryFrom PData (PAsData target)
+  , PIsData actual
+  , PIsData target
+  ) =>
+  ClosedTerm (PAsData actual) ->
+  ClosedTerm (PAsData target)
+checkDeep t = unTermCont $ fst <$> TermCont (ptryFrom @(PAsData target) $ pforgetData t)
```

Comment:
I was under the impression the checking is only for `PData`. Are we encouraging checking a correctly typed `PAsData` as well? If a `PAsData` is actually incorrect - doesn't that mean the typing has been compromised further up in a way it shouldn't have been?

Reply:
`PTryFrom` is generalised. You can e.g. have `PTryFrom (PAsData Integer) (PAsData Natural)`.

Reply:
ah, I see - we're overloading with any conversion, not just data ones. Cool

Reply:
although I do think the examples are a bit misleading for that purpose.
```hs
checkDeep
  @(PBuiltinPair (PAsData PInteger) (PAsData PByteString))
  @(PBuiltinPair (PAsData PByteString) (PAsData PByteString))
  (pdata $ ppairDataBuiltin # (pdata $ pconstant "foo") # (pdata $ pconstant "bar"))
```
I can't see this being idiomatic, but I might be missing something. The existence of such a check implies the `PBuiltinPair` typing is wrong and the types have been compromised somewhere.